R/AbstractGraphReporter.R

```diff
@@ -172,34 +172,122 @@ AbstractGraphReporter <- R6::R6Class(
-            private$nodes[, label := id]
+            plotDTnodes <- copy(private$nodes) # Don't modify original
+            plotDTnodes[, id := node]
+            plotDTnodes[, label := id]
             
-            private$edges[, from := SOURCE]
-            private$edges[, to := TARGET]
+            plotDTedges <- copy(private$edges) # Don't modify original
+            plotDTedges[, from := SOURCE]
+            plotDTedges[, to := TARGET]
+            plotDTedges[, color := '#848484'] # TODO Make edge formatting flexible too
             
-            defaultNodeColor <- "blue" 
-            if (is.null(colorFieldName)) {
+            
+            # Color By Field
+            if(is.null(private$plotNodeColorScheme[['field']])) {
+              
+              # Default Color for all Nodes
+              plotDTnodes[, color := private$plotNodeColorScheme[['pallete']]]
+              
+            } else {
+              colorFieldName <- private$plotNodeColorScheme[['field']]
+              colorFieldPallete <- private$plotNodeColorScheme[['pallete']]
+              colorFieldValues <- plotDTnodes[[colorFieldName]]
+              
+              log_info(sprintf("Coloring plot nodes by %s..."
+                               , colorFieldName))
+              
+              # If character field
+              if(is.character(colorFieldValues) | is.factor(colorFieldValues)) {
+                # Create pallete by unique values
+                valCount <- data.table::uniqueN(colorFieldValues)
+                newPallete <- grDevices::colorRampPalette(colors = colorFieldPallete)(valCount)
                 
-                # Same Color for all nodes
-                g <- visNetwork::visNetwork(nodes = private$nodes, edges = private$edges) %>%
-                     visNetwork::visHierarchicalLayout(sortMethod = "directed", direction = "DU") %>%
-                     visNetwork::visEdges(arrows = 'to') %>%
-                     visNetwork::visOptions(highlightNearest = list(enabled = TRUE, degree = 2000, algorithm = "hierarchical"))
                 
-            } else if (is.factor(private$nodes[, colorFieldName, with = FALSE]) | is.character(private$nodes[, colorFieldName, with = FALSE])) {
-                # Color By Discrete Factor
-            } else {
-                # Assume continuous number scale
-            }
+                # For each character value, update all nodes with that value
+                plotDTnodes[, color := newPallete[.GRP]
+                              , by = list(get(colorFieldName))]
+                
+              } else if (is.numeric(colorFieldValues)) {
+                # If numeric field, assume continuous
+                newPallete <- grDevices::colorRamp(colors = colorFieldPallete)
+                plotDTnodes[, color := newPallete(get(colorFieldName))]
+                
+              } else {
+                log_fatal("A character, factor, or numeric field can be used to color nodes.")
```

Comment:
can you please add another statement here? Like after saying "you can only use these types" we should be adding "column <column_name> is of type <whatever_type>"